asserts/pool.go

```diff
@@ -19,6 +19,538 @@
 
 package asserts
 
+import (
+	"errors"
+	"fmt"
+
+	"github.com/snapcore/snapd/asserts/internal"
+)
+
 // A Grouping identifies opaquely a grouping of assertions.
-// TODO: add details to this comment when this is actually used.
+// Pool uses it to label the interesection between a set of groups.
 type Grouping string
+
+// A pool helps holding and tracking a set of assertions and their
+// prerequisites as they need to be updated or resolved.  The
+// assertions can be organized in groups. Failure can be tracked
+// isolated to groups, conversely any error related to a single group
+// alone will stop any work to resolve it. Independent assertions
+// should not be grouped.  Assertions and prerequisites that are part
+// of more than one group are tracked properly only once.
+//
+// Typical usage involves specifying the initial assertions needing to
+// be resolved or updated using AddUnresolved and AddToUpdate. At this
+// point ToResolve can be called to get them organized in groupings
+// ready for fetching. Fetched assertions can then be provided with
+// Add or AddBatch. Because these can have prerequisites calling
+// ToResolve and fetching needs to be repeated until ToResolve's
+// result is empty.  Between any two ToResolve invocations but after
+// any Add or AddBatch AddUnresolved/AddToUpdate can also be used
+// again.
+//
+//                      V
+//                      |
+//        /-> AddUnresolved, AddToUpdate
+//        |             |
+//        |             V
+//        |------> ToResolve -> empty? done
+//        |             |
+//        |             V
+//        \ __________ Add
+//
+//  * TODO: AddToUpdate is not implemented yet
+//  * TODO: document Add*Error, CommitTo
+type Pool struct {
+	groundDB RODatabase
+
+	numbering map[string]uint16
+	groupings *internal.Groupings
+
+	unresolved    map[string]*unresolvedRec
+	prerequisites map[string]*unresolvedRec
+
+	bs     Backstore
+	groups map[uint16]*groupRec
+
+	curPhase poolPhase
+}
+
+// NewPool creates a new Pool, groundDB is used to resolve trusted and
+// predefined assertions and to provide the current revision for
+// assertions to update and their prerequisites. Up to n groups can be
+// used to organize the assertions.
+func NewPool(groundDB RODatabase, n int) *Pool {
+	groupings, err := internal.NewGroupings(n)
+	if err != nil {
+		panic(fmt.Sprintf("NewPool: %v", err))
```

Comment:
Where does n come from, is it guranteed to be sane for NewGroupings ctor? Since NewGroupings may report an error, could we make NewPool return (*Pool, err) too instead of panic?

Reply:
yes, is something fixed that one needs to pick, it doesn't come from input. At most is something we could chose based on memory of the machine if we bothered. Either way something that we control and shouldn't turn into an unexpected value at runtime.

Reply:
Ok, panic isn't an issue then, thanks for clarification.